File: swift/Sources/ForySwift/Fory.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Foundation

public struct ForyConfig {
    public var xlang: Bool
    public var trackRef: Bool
    public var compatible: Bool

    public init(xlang: Bool = true, trackRef: Bool = false, compatible: Bool = false) {
        self.xlang = xlang
        self.trackRef = trackRef
        self.compatible = compatible
    }
}

public final class Fory {
    public let config: ForyConfig
    public let typeResolver: TypeResolver

    public init(config: ForyConfig = ForyConfig()) {
        self.config = config
        self.typeResolver = TypeResolver()
    }

    public func register<T: Serializer>(_ type: T.Type, id: UInt32) {
        typeResolver.register(type, id: id)
    }

    public func register<T: Serializer>(_ type: T.Type, name: String) throws {
        try typeResolver.register(type, namespace: "", typeName: name)
    }

    public func register<T: Serializer>(_ type: T.Type, namespace: String, name: String) throws {
        try typeResolver.register(type, namespace: namespace, typeName: name)
    }

    public func serialize<T: Serializer>(_ value: T) throws -> Data {
        let writer = ByteWriter()
        writeHead(writer: writer, isNone: value.foryIsNone)

        if !value.foryIsNone {
            let compatibleTypeDefState = CompatibleTypeDefWriteState()
            let context = WriteContext(
                writer: writer,
                typeResolver: typeResolver,
                trackRef: config.trackRef,
                compatible: config.compatible,
                compatibleTypeDefState: compatibleTypeDefState,
                metaStringWriteState: MetaStringWriteState()
            )
            let refMode: RefMode = config.trackRef ? .tracking : .nullOnly
            try value.foryWrite(context, refMode: refMode, writeTypeInfo: true, hasGenerics: false)
            context.resetObjectState()
        }

        return writer.toData()
    }

    public func deserialize<T: Serializer>(_ data: Data, as _: T.Type = T.self) throws -> T {
        let reader = ByteReader(data: data)
        let isNone = try readHead(reader: reader)
        if isNone {
            return T.foryDefault()
        }

        let context = ReadContext(
            reader: reader,
            typeResolver: typeResolver,
            trackRef: config.trackRef,
            compatible: config.compatible,
            compatibleTypeDefState: CompatibleTypeDefReadState(),
            metaStringReadState: MetaStringReadState()
        )
        let refMode: RefMode = config.trackRef ? .tracking : .nullOnly
        let value = try T.foryRead(context, refMode: refMode, readTypeInfo: true)
        context.resetObjectState()
        return value
    }

    public func serializeTo<T: Serializer>(_ buffer: inout Data, value: T) throws {
        let writer = ByteWriter()
        writeHead(writer: writer, isNone: value.foryIsNone)
        if !value.foryIsNone {
            let context = WriteContext(
                writer: writer,
                typeResolver: typeResolver,
                trackRef: config.trackRef,
                compatible: config.compatible,
                compatibleTypeDefState: CompatibleTypeDefWriteState(),
                metaStringWriteState: MetaStringWriteState()
            )
            let refMode: RefMode = config.trackRef ? .tracking : .nullOnly
            try value.foryWrite(context, refMode: refMode, writeTypeInfo: true, hasGenerics: false)
            context.resetObjectState()
        }
        buffer.append(writer.toData())
    }

    public func deserializeFrom<T: Serializer>(_ reader: ByteReader, as _: T.Type = T.self) throws -> T {
        let isNone = try readHead(reader: reader)
        if isNone {
            return T.foryDefault()
        }
        let context = ReadContext(
            reader: reader,
            typeResolver: typeResolver,
            trackRef: config.trackRef,
            compatible: config.compatible,
            compatibleTypeDefState: CompatibleTypeDefReadState(),
            metaStringReadState: MetaStringReadState()
        )
        let refMode: RefMode = config.trackRef ? .tracking : .nullOnly
        let value = try T.foryRead(context, refMode: refMode, readTypeInfo: true)
        context.resetObjectState()
        return value
    }

    public func writeHead(writer: ByteWriter, isNone: Bool) {
        var bitmap: UInt8 = 0
        if config.xlang {
            bitmap |= ForyHeaderFlag.isXlang
        }
        if isNone {
            bitmap |= ForyHeaderFlag.isNull
        }
        writer.writeUInt8(bitmap)
    }

    public func readHead(reader: ByteReader) throws -> Bool {
        let bitmap = try reader.readUInt8()
        let peerIsXlang = (bitmap & ForyHeaderFlag.isXlang) != 0
        if peerIsXlang != config.xlang {
            throw ForyError.invalidData("xlang bitmap mismatch")
        }
        return (bitmap & ForyHeaderFlag.isNull) != 0
    }
}

File: csharp/src/Fory/Fory.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System.Buffers;

namespace Apache.Fory;

/// <summary>
/// Core serializer runtime.
/// This type is optimized for single-threaded reuse and must not be shared concurrently across threads.
/// Use <see cref="ThreadSafeFory"/> for concurrent access.
/// </summary>
public sealed class Fory
{
    private readonly TypeResolver _typeResolver;
    private WriteContext _writeContext;
    private ReadContext _readContext;

    internal Fory(Config config)
    {
        Config = config;
        _typeResolver = new TypeResolver();
        _writeContext = new WriteContext(
            new ByteWriter(),
            _typeResolver,
            Config.TrackRef,
            Config.Compatible,
            Config.CheckStructVersion,
            new CompatibleTypeDefWriteState(),
            new MetaStringWriteState());
        _readContext = new ReadContext(
            new ByteReader(Array.Empty<byte>()),
            _typeResolver,
            Config.TrackRef,
            Config.Compatible,
            Config.CheckStructVersion,
            new CompatibleTypeDefReadState(),
            new MetaStringReadState());
    }

    public Config Config { get; }

    public static ForyBuilder Builder()
    {
        return new ForyBuilder();
    }

    public Fory Register<T>(uint typeId)
    {
        _typeResolver.Register(typeof(T), typeId);
        return this;
    }

    public Fory Register<T>(string typeName)
    {
        _typeResolver.Register(typeof(T), string.Empty, typeName);
        return this;
    }

    public Fory Register<T>(string typeNamespace, string typeName)
    {
        _typeResolver.Register(typeof(T), typeNamespace, typeName);
        return this;
    }

    public Fory Register<T, TSerializer>(uint typeId)
        where TSerializer : Serializer<T>, new()
    {
        TypeInfo typeInfo = _typeResolver.RegisterSerializer<T, TSerializer>();
        _typeResolver.Register(typeof(T), typeId, typeInfo);
        return this;
    }

    public Fory Register<T, TSerializer>(string typeNamespace, string typeName)
        where TSerializer : Serializer<T>, new()
    {
        TypeInfo typeInfo = _typeResolver.RegisterSerializer<T, TSerializer>();
        _typeResolver.Register(typeof(T), typeNamespace, typeName, typeInfo);
        return this;
    }

    public byte[] Serialize<T>(in T value)
    {
        ByteWriter writer = _writeContext.Writer;
        writer.Reset();
        Serializer<T> serializer = _typeResolver.GetSerializer<T>();
        TypeInfo typeInfo = _typeResolver.GetTypeInfo<T>();
        bool isNone = typeInfo.IsNullableType && value is null;
        WriteHead(writer, isNone);
        if (!isNone)
        {
            _writeContext.ResetFor(writer);
            RefMode refMode = Config.TrackRef ? RefMode.Tracking : RefMode.NullOnly;
            serializer.Write(_writeContext, value, refMode, true, false);
            _writeContext.ResetObjectState();
        }

        return writer.ToArray();
    }

    public void Serialize<T>(IBufferWriter<byte> output, in T value)
    {
        byte[] payload = Serialize(value);
        output.Write(payload);
    }

    public T Deserialize<T>(ReadOnlySpan<byte> payload)
    {
        ByteReader reader = _readContext.Reader;
        reader.Reset(payload);
        T value = DeserializeFromReader<T>(reader);
        if (reader.Remaining != 0)
        {
            throw new InvalidDataException($"unexpected trailing bytes after deserializing {typeof(T)}");
        }

        return value;
    }

    public T Deserialize<T>(byte[] payload)
    {
        ByteReader reader = _readContext.Reader;
        reader.Reset(payload);
        T value = DeserializeFromReader<T>(reader);
        if (reader.Remaining != 0)
        {
            throw new InvalidDataException($"unexpected trailing bytes after deserializing {typeof(T)}");
        }

        return value;
    }

    public T Deserialize<T>(ref ReadOnlySequence<byte> payload)
    {
        byte[] bytes = payload.ToArray();
        ByteReader reader = _readContext.Reader;
        reader.Reset(bytes);
        T value = DeserializeFromReader<T>(reader);
        payload = payload.Slice(reader.Cursor);
        return value;
    }


    public void WriteHead(ByteWriter writer, bool isNone)
    {
        byte bitmap = 0;
        if (Config.Xlang)
        {
            bitmap |= ForyHeaderFlag.IsXlang;
        }

        if (isNone)
        {
            bitmap |= ForyHeaderFlag.IsNull;
        }

        writer.WriteUInt8(bitmap);
    }

    public bool ReadHead(ByteReader reader)
    {
        byte bitmap = reader.ReadUInt8();
        bool peerIsXlang = (bitmap & ForyHeaderFlag.IsXlang) != 0;
        if (peerIsXlang != Config.Xlang)
        {
            throw new InvalidDataException("xlang bitmap mismatch");
        }

        return (bitmap & ForyHeaderFlag.IsNull) != 0;
    }

    private T DeserializeFromReader<T>(ByteReader reader)
    {
        bool isNone = ReadHead(reader);
        Serializer<T> serializer = _typeResolver.GetSerializer<T>();
        if (isNone)
        {
            return serializer.DefaultValue;
        }

        _readContext.ResetFor(reader);
        RefMode refMode = Config.TrackRef ? RefMode.Tracking : RefMode.NullOnly;
        T value = serializer.Read(_readContext, refMode, true);
        _readContext.ResetObjectState();
        return value;
    }

}
